core/src/main/java/com/linecorp/armeria/server/healthcheck/ScheduledHealthChecker.java

```diff
@@ -0,0 +1,214 @@
+ * License for the specific language governing permissions and limitations
+ * under the Licenses
+ */
+
+package com.linecorp.armeria.server.healthcheck;
+
+import java.time.Duration;
+import java.util.concurrent.CompletionStage;
+import java.util.concurrent.Future;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.concurrent.atomic.AtomicReference;
+import java.util.function.Consumer;
+import java.util.function.Supplier;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.util.concurrent.Futures;
+
+import com.linecorp.armeria.common.util.AbstractListenable;
+
+import io.netty.util.concurrent.EventExecutor;
+
+/**
+ * A {@link ListenableHealthChecker} whose state can be set by the result of supplied CompletionStage.
+ * This class can be shared by multiple {@link HealthCheckService}.
+ */
+final class ScheduledHealthChecker extends AbstractListenable<HealthChecker>
+        implements ListenableHealthChecker {
+
+    private final Supplier<? extends CompletionStage<HealthCheckStatus>> healthChecker;
+    private final Duration fallbackTtl;
+    private final EventExecutor eventExecutor;
+    private final Consumer<HealthChecker> onHealthCheckerUpdate;
+    private final AtomicBoolean isHealthy;
+    private final AtomicInteger requestCount;
+    private final AtomicReference<ScheduledHealthCheckerImpl> impl;
+
+    ScheduledHealthChecker(Supplier<? extends CompletionStage<HealthCheckStatus>> healthChecker,
+                           Duration facllbackTtl, EventExecutor eventExecutor) {
+        this.healthChecker = healthChecker;
+        this.fallbackTtl = facllbackTtl;
+        this.eventExecutor = eventExecutor;
+
+        isHealthy = new AtomicBoolean(false);
+        requestCount = new AtomicInteger(0);
+        impl = new AtomicReference<>();
+        onHealthCheckerUpdate = latestValue -> {
+            isHealthy.set(latestValue.isHealthy());
+            notifyListeners(latestValue);
+        };
+    }
+
+    @Override
+    public boolean isHealthy() {
+        return isHealthy.get();
+    }
+
+    void startHealthChecker() {
+        if (requestCount.incrementAndGet() != 1) {
+            return;
+        }
+        final ScheduledHealthCheckerImpl newlyScheduled =
+                new ScheduledHealthCheckerImpl(healthChecker, fallbackTtl, eventExecutor);
+        for (;;) {
+            if (impl.compareAndSet(null, newlyScheduled)) {
+                newlyScheduled.addListener(onHealthCheckerUpdate);
+                newlyScheduled.startHealthChecker();
+                break;
+            }
+        }
+    }
+
+    void stopHealthChecker() {
+        final int previousCount = requestCount.getAndUpdate(count -> {
+            if (count > 0) {
+                return count - 1;
+            }
+            return 0;
+        });
```

Comment:
When do we call `stopHealthChecker()` more then `startHealthChecker()`? Otherwise we could just use `getAndDecrease()`.

Reply:
right, I just try to make it safe by it self. It's better to throw exception if count < 0?

Reply:
Yeah, how about assertion?